### PR TITLE
Improve type checking across codebase

### LINKED ISSUE #55: Improve compiler validations to support use as library
As `pact-node` is a library (in addition to a tool) and will be embedded in potentially many others, we have a duty to ensure the highest levels of TS language use.

An annoying restriction of TS is that it will compile, and therefore complain, about upstream dependencies.

e.g. for `pact-js` even with exclusions set as follows:

```
  "include": [
    "./src/**/*"
  ],
  "exclude": [
    "./node_modules/**/*",
    "./dist",
    "./examples",
    "**/*.spec.ts"
  ]
```

TS still picts up `pact-node`, as it is in compilation scope, so all compiler options for the external project are applied equally to `pact-node`, resulting in lots of errors (I disabled the `noImplicitAny` to reduce noise):

```
node_modules/@pact-foundation/pact-node/src/broker.ts(70,21): error TS2532: Object is possibly 'undefined'.
node_modules/@pact-foundation/pact-node/src/broker.ts(70,53): error TS2532: Object is possibly 'undefined'.
node_modules/@pact-foundation/pact-node/src/broker.ts(81,17): error TS2345: Argument of type 'any' is not assignable to parameter of type 'never'.
node_modules/@pact-foundation/pact-node/src/logger.ts(11,8): error TS2339: Property 'info' does not exist on type 'Logger'.
node_modules/@pact-foundation/pact-node/src/logger.ts(19,36): error TS2339: Property 'level' does not exist on type 'Logger'.
node_modules/@pact-foundation/pact-node/src/logger.ts(25,16): error TS2554: Expected 0 arguments, but got 1.
node_modules/@pact-foundation/pact-node/src/pact-util.ts(7,9): error TS7030: Not all code paths return a value.
node_modules/@pact-foundation/pact-node/src/publisher.ts(100,20): error TS2345: Argument of type 'any' is not assignable to parameter of type 'never'.
node_modules/@pact-foundation/pact-node/src/server.ts(243,4): error TS2322: Type 'undefined' is not assignable to type 'ChildProcess'.
node_modules/@pact-foundation/pact-node/src/verifier.ts(31,25): error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.
node_modules/@pact-foundation/pact-node/src/verifier.ts(112,7): error TS2532: Object is possibly 'undefined'.
src/dsl/verifier.ts(8,25): error TS2339: Property 'verifyPacts' does not exist on type 'typeof "/Users/mfellows/development/public/pact-js/node_modules/@pact-foundation/pact-node/src/in...'.
src/pact.ts(59,34): error TS2339: Property 'createServer' does not exist on type 'typeof "/Users/mfellows/development/public/pact-js/node_modules/@pact-foundation/pact-node/src/in...'.
src/pact.ts(70,20): error TS2339: Property 'logLevel' does not exist on type 'typeof "/Users/mfellows/development/public/pact-js/node_modules/@pact-foundation/pact-node/src/in...'.
```

My suggestion is that we increase strictness/validation compilation rules, but I'm open to other alternatives.

### PR DESCRIPTION
**Pull request by mefellows**
Adds the following TS configuration, in order to improve the libraries portability into other TS projects:

* --noImplicitAny: Raise error on expressions and declarations with an implied any type.
* --noImplicitReturns: Report error when not all code paths in function return a value.
* --noImplicitThis: Raise error on this expressions with an implied any type.
* --strictNullChecks: In strict null checking mode, the nulland undefined values are not in the domain of every type and are only assignable to themselves and any (the one exception being that undefined is also assignable to void).

NOTE: in a few q-related cases, explicit type casting was necessary. I haven't been able to dig far enough into why the code works but types are incorrect - either there is an issue with the code, or the types aren't quite correct (in particular, a `q.Resolve(...)` is not compatible with the `PromiseLike` interface).

Fixes #55.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -14,3 +14,7 @@ trim_trailing_whitespace = false
 [*.js]
 indent_style = tab
 indent_size = 4
+
+[package.json]
```
Comment (mefellows) [inline .editorconfig]: This should prevent that reformatting ;)

File: src/broker.ts [hunk 2]
```diff
@@ -51,32 +51,32 @@ export class Broker {
...
 		return q.all(promises)
-			.then((values) => _.reduce(values, (array, v) => {
+			.then((values) => _.reduce(values, (array: string[], v: any) => {
 				if (v && v._links && v._links.pacts) {
 					array.push(..._.pluck(v._links.pacts, "href"));
 				}
 				return array;
 			}, []))
-			.catch(() => q.reject(`Unable to find pacts for given provider '${this.options.provider}' and tags '${this.options.tags}'`));
+			.catch(() => q.reject(`Unable to find pacts for given provider '${this.options.provider}' and tags '${this.options.tags}'`) as IWhenable<any>);
```
Comment (mefellows) [inline src/broker.ts]: This is an example of that `q` problem - a `q.Promise` is not compatible with a `IWhenable<T>` which it thinks a `q.all(...)` returns. 

Another potential option here is to modify the `q` import to not import the types, but you do lose out on some of the nice code completion and type completeness elsewhere in the file.

**Comment by mefellows** [comment]
Any thoughts @mboudreau or should we merge this in (no API compatibility changes).

**Comment by mboudreau** [comment]
@mefellows Ah, right, I'll look at this today.

File: src/broker.ts (hunk 2 above)
Comment (mboudreau) [reply src/broker.ts]: isn't that the reason why q support generics?

File: bin/pact-cli.spec.ts [hunk 3]
```diff
@@ -3,11 +3,11 @@ import chaiAsPromised = require("chai-as-promised");
 import childProcess = require("child_process");
 import q = require("q");
 import path = require("path");
-import _ = require("underscore");
 import {ChildProcess} from "child_process";
 import {ServerOptions} from "../src/server";
-import decamelize = require("decamelize");
 import providerMock from "../test/integration/provider-mock";
+const decamelize = require("decamelize");
```
Comment (mboudreau) [inline bin/pact-cli.spec.ts]: isn't import and const the same thing in this case?

File: src/pact.spec.ts [hunk 4]
```diff
@@ -5,14 +5,14 @@ import fs = require("fs");
...
-
 chai.use(chaiAsPromised);
+const currentDir = (process && process.mainModule) ? process.mainModule.filename : "";
 
 describe("Pact Spec", () => {
 	afterEach(() => pact.removeAllServers());
 
 	describe("Set Log Level", () => {
-		let originalLogLevel;
+		let originalLogLevel: any;
```
Comment (mboudreau) [inline src/pact.spec.ts]: string

File: src/pact.spec.ts [hunk 5]
```diff
@@ -51,7 +51,7 @@ describe("Pact Spec", () => {
 	});
 
 	describe("Create serverFactory", () => {
-		let dirPath;
+		let dirPath: any;
```
Comment (mboudreau) [inline src/pact.spec.ts]: string

File: src/pact.ts [hunk 6]
```diff
@@ -32,7 +32,7 @@ export class Pact {
 		logger.info(`Creating Pact Server with options: \n${this.__stringifyOptions(server.options)}`);
 
 		// Listen to server delete events, to remove from server list
-		server.once("delete", (s) => {
+		server.once("delete", (s: any) => {
```
Comment (mboudreau) [inline src/pact.ts]: Pact

File: src/server.spec.ts [hunk 7]
```diff
@@ -12,7 +12,7 @@ chai.use(chaiAsPromised);
 const expect = chai.expect;
 
 describe("Server Spec", () => {
-	let server;
+	let server: any;
```
Comment (mboudreau) [inline src/server.spec.ts]: Pact

File: src/server.spec.ts [hunk 8]
```diff
@@ -57,7 +57,7 @@ describe("Server Spec", () => {
 		});
 
 		context("when valid options are set", () => {
-			let dirPath;
+			let dirPath: any;
```
Comment (mboudreau) [inline src/server.spec.ts]: string

Comment (mefellows) [reply src/server.spec.ts]: 👍

File: src/server.spec.ts (hunk 7 above)
Comment (mefellows) [reply src/server.spec.ts]: I think you mean `Server`. Fixing.

File: src/pact.spec.ts (hunk 5 above)
Comment (mefellows) [reply src/pact.spec.ts]: 👍

File: src/pact.spec.ts (hunk 4 above)
Comment (mefellows) [reply src/pact.spec.ts]: Actually it's a union type of `LogLevels` and `number | void`, as it gets use to both set and retrieve the log level - hence why I put any.

File: bin/pact-cli.spec.ts (hunk 3 above)
Comment (mefellows) [reply bin/pact-cli.spec.ts]: No, `const` doesn't try to resolve the module whereas `import` does. e.g. in this case the following error appears when using `import`:

```sh
bin/pact-cli.spec.ts(9,29): error TS7016: Could not find a declaration file for module 'decamelize'. '/Users/mfellows/development/public/pact-node/node_modules/decamelize/index.js' implicitly has an 'any' type.
  Try `npm install @types/decamelize` if it exists or add a new declaration (.d.ts) file containing `declare module 'decamelize';`
```

**Pull request merged**